module3/lessons/error_handling_review.md

```diff
@@ -0,0 +1,322 @@
+---
+layout: page
+title: Error Handling in Rails Applicaitons
+length: 120
+tags: rails, errors, rescue, raise, exceptions
+---
+
+## Vocabulary
+* Exception
+* Error
+* Class Hierarchy
+* Rescue
+
+## Learning Goals
+
+* Understand the need as developers to plan for errors and handle exceptions
+* Review an object oriented approach to error handling
+
+### Setup
+
+We'll be using the `error_handling_start` branch of [this repository](https://github.com/turingschool-examples/house-salad-7) to review error handling and exceptions in Rails. The completed code can be found on the `error_handling_complete` branch.
+
+Our Facade currently looks like so:
+
+```ruby
+class SearchFacade
+  def initialize(state = nil, last_name = nil)
+    @state = state
+    @last_name = last_name
+  end
+
+  def all_senate_members
+    json = CongressService.new.senate_members[:results][0][:members]
+  end
+
+  def senate_member_by_last_name
+    match = members_by_last_name(@last_name).first
+    SenateMember.new(member)
+  end
+
+  def members_by_last_name(last_name)
+    all_senate_members.find_all do |member|
+      member[:last_name] == last_name
+    end
+  end
+
+  def house_members
+    service = CongressService.new
+
+    json = service.members_by_state(@state)
+    
+    @members = json[:results].map do |member_data|
+      Member.new(member_data)
+    end
+  end
+end
+```
+
+First, lets isolate our call to our service.
+
+```ruby
+def senate_service
+  @@service_result ||= CongressService.senate_members
+end
+
+def all_senate_members
+  senate_service[:results][0][:members]
+end
+```
+Isolation helps on a few fronts. It adheres more closely to SRP, and we can properly test this method. We can also take advantage of memoization to ensure our we only ever make one external API call per request.
+
+Because we are memomizing within a class method, we need a class variable to hold the result of our service call. An instance variable or local variable will not memoize properly.
```

Comment:
Typo `memomizing` => `memoizing`

Reply:
I feel like memoization could be a whole other topic honestly, and while I like that we introduce it, we're also introducing the concept of a class variable which we haven't done before. Still thinking about this, but I would almost want to link to Rails/Ruby docs with these topics, or make another sub-lesson that explains them deeper. 🤔

Reply:
Do we also have a completed version of this review on a branch in house-salad-7? If so, can we link it at the bottom under the Checks for Understanding?

Reply:
1) I agree-- after working through this review I don't think these examples of memoization and class variables are valuable, anyway. Especially the memoization in `members_by_last_name`, which can actually cause a result to be returned regardless of whether the user's query matches any data. I'll take out the references to class vars and memoization
2) Yes, there's a completed version on `error_handling_complete`, I'll be sure to add that at the end of the lesson.